libcudacxx/include/cuda/__iterator/permutation_iterator.h

```diff
@@ -0,0 +1,435 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___ITERATOR_PERMUTATION_ITERATOR_H
+#define _CUDA___ITERATOR_PERMUTATION_ITERATOR_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__concepts/equality_comparable.h>
+#include <cuda/std/__concepts/totally_ordered.h>
+#include <cuda/std/__iterator/concepts.h>
+#include <cuda/std/__iterator/iterator_traits.h>
+#include <cuda/std/__type_traits/is_nothrow_copy_constructible.h>
+#include <cuda/std/__type_traits/is_nothrow_move_constructible.h>
+#include <cuda/std/__utility/declval.h>
+#include <cuda/std/__utility/move.h>
+
+#if _LIBCUDACXX_HAS_SPACESHIP_OPERATOR()
+#  include <cuda/std/detail/libcxx/include/compare>
+#endif // _LIBCUDACXX_HAS_SPACESHIP_OPERATOR()
+
+#include <cuda/std/__cccl/prologue.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+//! @p permutation_iterator is an iterator which represents a pointer into a reordered view of a given range. \p
+//! permutation_iterator is an imprecise name; the reordered view need not be a strict permutation. This iterator is
+//! useful for fusing a scatter or gather operation with other algorithms.
+//!
+//! This iterator takes two arguments:
+//!
+//!   - an iterator to the range \c V on which the "permutation" will be applied
+//!   - an iterator to a range of indices defining the reindexing scheme that determines how the elements of \c V will
+//!   be permuted.
+//!
+//! Note that \p permutation_iterator is not limited to strict permutations of the given range \c V. The distance
+//! between begin and end of the reindexing iterators is allowed to be smaller compared to the size of the range \c V,
```

Comment:
@pauleonix Here we forbid the resulting range to be larger than the base range, so we should not change `ìterator_difference`

The issue I have with any modification here is that it might actually break user code.

Consider a user iterator that for some performance reason uses a smaller `difference_type`. If we would change `permutation_iterator` to allow the new range to be larger, than this would change the size of the resulting iterator irrespective of whether the user actually wants this feature

Reply:
Correct me if I'm wrong, but I think the Boost/Thrust implementation does allow bigger ranges and does use the `difference_type` from the offset/index iterator. So this feels like a significant restriction (even though in practice assuming both use `ptrdiff_t` users could still use a bigger resulting range). Another possibly confusing difference is that in Thrust the `base` is actually the offset/index iterator. For another use case see https://stackoverflow.com/q/74190787

Reply:
We semantically forbid that

> The distance between begin and end of the reindexing iterators is allowed to be smaller compared to the size of the range V

Note there there is no way an iterator could assert the size of the range it builds